### PR TITLE
fix(ItemsControl): desync from modifying ItemsSource of non-INCC/IObservableVector source

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): closes unoplatform/nventive-private#457, closes unoplatform/nventive-private#472

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?
When ListView.ItemsSource is assigned/bound to a non-observable/notifying collection, like List<T>,
adding/removing item to the collection will cause the LV to update and the items' data-context to desync.

## What is the new behavior?
Like on UWP, adding/removing item to the non-observable/notifying collection will NOT update the LV.

## PR Checklist

Please check if your PR fulfills the following requirements:
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by Xiaoy312** [comment]
saving top comment

File: src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs [hunk 1]
```diff
@@ -671,7 +673,33 @@ private void TryObserveCollectionViewSource(object newValue)
 		// CollectionViewSource.View to whatever is being bound to (ie the ItemsSource). This should be fixed at some point because it
 		// has observable consequences in some usages.
 		internal object UnwrapItemsSource()
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:675]: this method is likely a hot path;
should we optimize it by:
- adding early-exit/fast-path for known types first?
- memorizing `IsGenericDescentOf` per `source.GetType()`?

or it matters little?

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions.Compatibility/Reflection/TypeExtensions.cs [hunk 2]
```diff
@@ -187,5 +189,40 @@ public static bool IsNullable(this Type type)
 				|| Nullable.GetUnderlyingType(type) != null;    // is Nullable<T>
 		}
 #endif
+
+		public static bool IsGenericDescentOf(this Type type, Type genericTypeDefinition)
```
Comment (Xiaoy312) [inline src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions.Compatibility/Reflection/TypeExtensions.cs:193]: drop commented code

File: src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs [hunk 3]
```diff
@@ -671,7 +673,33 @@ private void TryObserveCollectionViewSource(object newValue)
...
+				return cvs.View;
+			}
+
+			// For normal enumerables, that are not notifying (INCC) or observable (ObsCollection),
+			// any further change on the original source should not affect the rendering.
+			// Therefore, we want to use a shallow copy here instead of the original source
+			// to avoid changes being synchronized onto the materialized items on subsequent measure calls.
+			if (source is IEnumerable enumerable &&
+				source is not INotifyCollectionChanged or IObservableVector &&
+				!source.GetType().IsGenericDescentOf(typeof(IObservableVector<>)))
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:691]: This looks like an expensive operation. Should we be caching that operation?

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:691]: Ah you noted that above. Yes we should :)

File: src/Uno.UI.RuntimeTests/Helpers/ViewModelBase.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using System.Text;
+using System.Runtime.CompilerServices;
+
+namespace Uno.UI.RuntimeTests.Helpers
+{
+	public class ViewModelBase : INotifyPropertyChanged
+	{
+		public event PropertyChangedEventHandler PropertyChanged;
+
+		protected void RaiseAndSetIfChanged<T>(ref T backingField, T value, [CallerMemberName] string propertyName = null)
```
Comment (dr1rrb) [inline src/Uno.UI.RuntimeTests/Helpers/ViewModelBase.cs:13]:
```suggestion
		protected void SetAndRaiseIfChanged<T>(ref T backingField, T value, [CallerMemberName] string propertyName = null)
```

:)

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions.Compatibility/Reflection/TypeExtensions.cs [hunk 5]
```diff
@@ -187,5 +189,37 @@ public static bool IsNullable(this Type type)
...
```
Comment (dr1rrb) [inline src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions.Compatibility/Reflection/TypeExtensions.cs:193]: Please add comment on those kind of method. For instance should the `genericTypeDefinition` has to be 'constructed' (having arguments for the generic parameters)?

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs [hunk 6]
```diff
@@ -2778,330 +2729,356 @@ public async Task When_Smooth_Scrolling()
 #endif
 
 		[TestMethod]
-		public async Task When_SelectionMode_Is_Multiple()
```
Comment (dr1rrb) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:2781]: Why did you removed this test? Is it expected?

Comment (Xiaoy312) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:2781]: good catch, no. must be copy-pasted over

File: src/Uno.UI.RuntimeTests/Helpers/ViewModelBase.cs (hunk 4 above)
Comment (Xiaoy312) [reply src/Uno.UI.RuntimeTests/Helpers/ViewModelBase.cs:13]: the name is actually taken from reactiveui's ReactiveObject.RaiseAndSetIfChanged, or inspired from
but SetAndRaiseIfChanged is okay too

**Comment by jeromelaban** [comment]
@mergifyio backport release/stable/4.8

**Pull request merged**